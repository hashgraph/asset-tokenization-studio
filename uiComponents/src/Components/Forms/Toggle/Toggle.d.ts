import type { switchAnatomy as ChakraSwitchParts } from "@chakra-ui/anatomy";
import type { SwitchProps as ChakraSwitchProps } from "@chakra-ui/switch";
import type { BaseMultiStyleConfiguration } from "@/Theme/Components/BaseMultiStyleConfiguration";
import type { ComponentWithAs } from "@chakra-ui/system";
export declare const togglePartsList: Array<(typeof ChakraSwitchParts.keys)[number] | "label">;
type Parts = typeof togglePartsList;
export interface ToggleThemeConfiguration extends BaseMultiStyleConfiguration<Parts> {
}
export interface ToggleProps extends Omit<ChakraSwitchProps, "placeholder"> {
    label?: string;
}
export declare const Toggle: ComponentWithAs<"input", ToggleProps>;
export {};
