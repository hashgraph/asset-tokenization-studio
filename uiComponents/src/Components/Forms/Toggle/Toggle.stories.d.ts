import React from "react";
declare const meta: import("@storybook/types").ComponentAnnotations<import("@storybook/react/dist/types-0a347bb9").R, import("@chakra-ui/system/dist/system.types").MergeWithAs<React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>, any, import("./Toggle").ToggleProps, import("@chakra-ui/system/dist/system.types").As>>;
export default meta;
export declare const Simple: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0a347bb9").R, import("@chakra-ui/system/dist/system.types").MergeWithAs<React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>, any, import("./Toggle").ToggleProps, import("@chakra-ui/system/dist/system.types").As>>;
export declare const IsDisabled: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0a347bb9").R, import("@chakra-ui/system/dist/system.types").MergeWithAs<React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>, any, import("./Toggle").ToggleProps, import("@chakra-ui/system/dist/system.types").As>>;
export declare const IsInvalid: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0a347bb9").R, import("@chakra-ui/system/dist/system.types").MergeWithAs<React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>, any, import("./Toggle").ToggleProps, import("@chakra-ui/system/dist/system.types").As>>;
export declare const Compound: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0a347bb9").R, import("@chakra-ui/system/dist/system.types").MergeWithAs<React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>, any, import("./Toggle").ToggleProps, import("@chakra-ui/system/dist/system.types").As>>;
export declare const ControlledToggle: () => JSX.Element;
