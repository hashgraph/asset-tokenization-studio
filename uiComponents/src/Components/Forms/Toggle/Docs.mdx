import {
  Story,
  Meta,
  Title,
  ArgTypes,
  Primary,
  Stories,
  Source,
} from "@storybook/addon-docs";
import { Grid, GridItem, Text, Link, Center, Flex } from "@chakra-ui/react";
import * as ToggleStories from "./Toggle.stories";

<Meta of={ToggleStories} />

<Title />

<a
  href="https://www.figma.com/file/YF8AvIq0MclfhGDvKgzypp/ioBricks-Design-System?type=design&node-id=1874%3A15429&mode=design&t=y5pso0dpnreGv6ac-1"
  target="\_blank"
>
  View figma design
</a>

<Primary />

## How to set up

### Config theme

Add the following configurations to the `theme.tsx` and `Toggle.ts` files, respectively.

```tsx dark
const theme = {
  components: {
    Switch: ConfigToggle,
  },
};
```

```tsx dark
const ConfigToggle: ToggleThemeConfiguration = {
  parts: togglePartsList,
  baseStyle: {},
};
```

### Component parts

You have these parts available: `"container" | "track" | "thumb" | "label"` (`togglePartsList`)

## Props

<ArgTypes of={ToggleStories} />

## Stories

<Stories title="" />
