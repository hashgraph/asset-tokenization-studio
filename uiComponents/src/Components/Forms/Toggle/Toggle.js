var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { Flex as ChakraFlex, FormLabel as ChakraFormLabel, } from "@chakra-ui/react";
import { Switch as ChakraSwitch } from "@chakra-ui/switch";
import { forwardRef, useMultiStyleConfig as useChakraMultiStyleConfig, } from "@chakra-ui/system";
import React from "react";
export var togglePartsList = ["container", "track", "thumb", "label"];
export var Toggle = forwardRef(function (_a, ref) {
    var label = _a.label, props = __rest(_a, ["label"]);
    var styles = useChakraMultiStyleConfig("Switch", {
        size: props.size,
        variant: props.variant
    });
    return (React.createElement(ChakraFlex, { align: "center" },
        React.createElement(ChakraSwitch, __assign({ ref: ref }, props)),
        React.createElement(ChakraFormLabel, { htmlFor: props.id, sx: styles.label }, label)));
});
