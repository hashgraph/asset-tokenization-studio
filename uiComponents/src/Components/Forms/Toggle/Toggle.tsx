import type { switchAnatomy as ChakraSwitchParts } from "@chakra-ui/anatomy";

import {
  Flex as ChakraFlex,
  FormLabel as ChakraFormLabel,
} from "@chakra-ui/react";
import type { SwitchProps as ChakraSwitchProps } from "@chakra-ui/switch";
import { Switch as ChakraSwitch } from "@chakra-ui/switch";
import {
  forwardRef,
  useMultiStyleConfig as useChakraMultiStyleConfig,
} from "@chakra-ui/system";
import React from "react";
import type { BaseMultiStyleConfiguration } from "@/Theme/Components/BaseMultiStyleConfiguration";
import type { ComponentWithAs } from "@chakra-ui/system";

export const togglePartsList: Array<
  (typeof ChakraSwitchParts.keys)[number] | "label"
> = ["container", "track", "thumb", "label"];

type Parts = typeof togglePartsList;

export interface ToggleThemeConfiguration
  extends BaseMultiStyleConfiguration<Parts> {}

export interface ToggleProps extends Omit<ChakraSwitchProps, "placeholder"> {
  label?: string;
}

export const Toggle: ComponentWithAs<"input", ToggleProps> = forwardRef<
  ToggleProps,
  "input"
>(({ label, ...props }: ToggleProps, ref) => {
  const styles = useChakraMultiStyleConfig("Switch", {
    size: props.size,
    variant: props.variant,
  });
  return (
    <ChakraFlex align="center">
      <ChakraSwitch ref={ref} {...props} />
      <ChakraFormLabel htmlFor={props.id} sx={styles.label}>
        {label}
      </ChakraFormLabel>
    </ChakraFlex>
  );
});
