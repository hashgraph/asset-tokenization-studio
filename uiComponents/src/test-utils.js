import { ChakraProvider, extendTheme } from "@chakra-ui/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";
import { BasePlatformTheme } from "./Theme/theme";
var DefaultTheme = extendTheme(BasePlatformTheme);
export var render = function (ui, theme) {
    if (theme === void 0) { theme = DefaultTheme; }
    var wrapper = function (_a) {
        var children = _a.children;
        return (React.createElement(ChakraProvider, { theme: theme }, children));
    };
    return testingLibraryRender(ui, { wrapper: wrapper });
};
