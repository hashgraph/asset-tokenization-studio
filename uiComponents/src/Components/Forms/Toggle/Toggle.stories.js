var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { Button } from "@Components/Interaction/Button";
import { linkTo } from "@storybook/addon-links";
import React from "react";
import { Toggle } from "./Toggle";
var meta = {
    title: "Design System/Forms/Toggle",
    component: Toggle,
    argTypes: {
        isDisabled: {
            control: { type: "boolean" },
            description: "Boolean to specify if the input is disabled."
        },
        isInvalid: {
            control: { type: "boolean" },
            description: "Boolean to specify if the input is invalid."
        },
        size: {
            control: false,
            description: "The size of the toggle. Must be defined in the theme (inside Switch component)"
        },
        variant: {
            description: "The variant of the toggle. Must be defined in the theme (inside Switch component)"
        },
        label: { description: "The label of the toggle" }
    },
    parameters: {
        design: {
            type: "figma",
            url: "https://www.figma.com/file/YF8AvIq0MclfhGDvKgzypp/NEW---ioBricks-Library?node-id=1874%3A15429&t=GPCIOt6k4o0cCWIe-4"
        },
        docs: {}
    },
    args: {
        size: "md"
    }
};
export default meta;
var Template = function (args) {
    return React.createElement(Toggle, __assign({}, args));
};
export var Simple = Template.bind({});
Simple.args = {};
export var IsDisabled = Template.bind({});
IsDisabled.args = {
    isDisabled: true
};
export var IsInvalid = Template.bind({});
IsInvalid.args = {
    isInvalid: true
};
export var Compound = Template.bind({});
Compound.args = {
    label: "Click to Toggle! :)"
};
export var ControlledToggle = function () { return (React.createElement(Button, { onClick: linkTo("Design System/Forms/Controllers/ToggleController") }, "Check out the ToggleController component Stories")); };
