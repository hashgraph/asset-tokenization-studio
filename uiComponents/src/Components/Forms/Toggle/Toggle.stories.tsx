import { Button } from "@Components/Interaction/Button";

import { linkTo } from "@storybook/addon-links";
import type { Meta, StoryFn } from "@storybook/react";
import React from "react";
import { Toggle } from "./Toggle";

const meta = {
  title: "Design System/Forms/Toggle",
  component: Toggle,
  argTypes: {
    isDisabled: {
      control: { type: "boolean" },
      description: "Boolean to specify if the input is disabled.",
    },
    isInvalid: {
      control: { type: "boolean" },
      description: "Boolean to specify if the input is invalid.",
    },
    size: {
      control: false,
      description:
        "The size of the toggle. Must be defined in the theme (inside Switch component)",
    },
    variant: {
      description:
        "The variant of the toggle. Must be defined in the theme (inside Switch component)",
    },
    label: { description: "The label of the toggle" },
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/YF8AvIq0MclfhGDvKgzypp/NEW---ioBricks-Library?node-id=1874%3A15429&t=GPCIOt6k4o0cCWIe-4",
    },
    docs: {},
  },
  args: {
    size: "md",
  },
} as Meta<typeof Toggle>;
export default meta;

const Template: StoryFn<typeof Toggle> = (args) => {
  return <Toggle {...args} />;
};

export const Simple = Template.bind({});
Simple.args = {};

export const IsDisabled = Template.bind({});
IsDisabled.args = {
  isDisabled: true,
};

export const IsInvalid = Template.bind({});
IsInvalid.args = {
  isInvalid: true,
};

export const Compound = Template.bind({});
Compound.args = {
  label: "Click to Toggle! :)",
};

export const ControlledToggle = () => (
  <Button onClick={linkTo("Design System/Forms/Controllers/ToggleController")}>
    Check out the ToggleController component Stories
  </Button>
);
