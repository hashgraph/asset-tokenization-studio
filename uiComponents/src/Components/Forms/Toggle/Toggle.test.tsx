import { render } from "@/test-utils";
import type { RenderResult } from "@testing-library/react";
import { waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import type { ToggleProps } from "./Toggle";
import { Toggle } from "./Toggle";

describe(`<Toggle />`, () => {
  const testId = "radioTest";

  const defaultProps: ToggleProps = {};

  const factoryComponent = (props = defaultProps) =>
    render(<Toggle data-testid={testId} {...props} />);

  const getInput = (component: RenderResult) =>
    component.getByTestId(testId).querySelector("input");

  test("renders correctly", () => {
    const component = factoryComponent();

    expect(component.asFragment()).toMatchSnapshot();
  });

  test("could be marked as disabled", () => {
    const component = factoryComponent({ isDisabled: true });

    expect(getInput(component)).toBeDisabled();
  });

  test("could set value", () => {
    const component = factoryComponent({ isChecked: true });

    expect(getInput(component)).toBeChecked();
  });

  test("value could be changed", async () => {
    const onChangeCallback = jest.fn();
    const component = factoryComponent({ onChange: onChangeCallback });

    expect(getInput(component)).not.toBeChecked();

    await userEvent.click(component.getByTestId(testId));

    await waitFor(() => {
      expect(onChangeCallback).toHaveBeenCalled();
      expect(getInput(component)).toBeChecked();
    });
  });
});
