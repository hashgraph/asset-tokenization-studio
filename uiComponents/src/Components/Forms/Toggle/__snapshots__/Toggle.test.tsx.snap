// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toggle /> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-70qvj9"
  >
    <label
      class="chakra-switch css-ghot30"
      data-testid="radioTest"
    >
      <input
        class="chakra-switch__input"
        style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
        type="checkbox"
        value=""
      />
      <span
        aria-hidden="true"
        class="chakra-switch__track css-1txbgvp"
      >
        <span
          class="chakra-switch__thumb css-wmswoc"
        />
      </span>
    </label>
    <label
      class="chakra-form__label css-1ii0m7i"
    />
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
